File: nl_banadir/public/js/purchase_order.js

frappe.ui.form.on("Purchase Order Item",{

    item_code: function(frm, cdt, cdn){
      
        var child = locals[cdt][cdn];
        if(child.custom_work_order){
        frappe.call({
            method: "nl_banadir.banadir_customization_reports.controllers.purchase_order.get_qty_from_first_work_order",
            args: {
                production_plan: frm.doc.custom_production_plan,
                work_order: child.custom_work_order,
            },
            callback: function(response){
                if(response.message){
                    frappe.model.set_value(cdt, cdn, "fg_item_qty", response.message.work_order_qty);
                    frappe.model.set_value(cdt, cdn, "fg_item", response.message.upper_stock_items);


                }
            }
        });
      }
    
    },
    custom_work_order: function(frm, cdt, cdn){
      
        var child = locals[cdt][cdn];
        frappe.call({
            method: "nl_banadir.banadir_customization_reports.controllers.purchase_order.get_qty_from_first_work_order",
            args: {
                production_plan: frm.doc.custom_production_plan,
                work_order: child.custom_work_order,
            },
            callback: function(response){
                if(response.message){
                    frappe.model.set_value(cdt, cdn, "fg_item_qty", response.message.work_order_qty);
                    frappe.model.set_value(cdt, cdn, "fg_item", response.message.upper_stock_items);


                }
            }
        });
      
    }
})


frappe.ui.form.on("Purchase Order", {
    refresh: function (frm) {
      frm.add_custom_button(
        __("Production Plan"),
        function () {
          if (!frm.doc.custom_production_plan) {
            frappe.msgprint(__("Please select a Production Plan first."));
            return;
          }
  
          frappe.call({
            method: "nl_banadir.banadir_customization_reports.controllers.purchase_order.get_items_from_production_plan",
            args: {
              production_plan: frm.doc.custom_production_plan,
            },
            callback: function (response) {
              if (response.message) {
                const items = response.message;
                frm.clear_table("items");
                items.forEach((item) => {
                  const child = frm.add_child("items"); // Replace "items" with your child table fieldname
                  frappe.model.set_value(child.doctype, child.name, "item_code", "JOB WORK CHARGES");
                  frappe.model.set_value(child.doctype, child.name, "fg_item", item.item_code);
                  frappe.model.set_value(child.doctype, child.name, "fg_item_qty", item.fg_item_qty);
                  frappe.model.set_value(child.doctype, child.name, "qty", item.quantity);
                  frappe.model.set_value(child.doctype, child.name, "custom_work_order", item.work_order);
                  frappe.model.set_value(child.doctype, child.name, "uom", "Pair");
                  frappe.model.set_value(child.doctype, child.name, "production_plan", frm.doc.custom_production_plan);
                });
  
                frm.refresh_field("items");
                frappe.msgprint(__("Items added from Production Plan."));
              }
            },
          });
        },
        __("Get Items From")
      );
    },
  });
  
  

File: nl_banadir/banadir_customization_reports/controllers/purchase_order.py
import frappe

@frappe.whitelist()
def get_items_from_production_plan(production_plan):
    """
    Fetch items and quantities from all Work Orders associated with the given Production Plan.
    """
    work_orders = frappe.get_all(
        "Work Order",
        filters={"production_plan": production_plan},
        fields=["name", "production_item", "qty"]
    )

    items = []
    for work_order in work_orders:
        items.append({
            "item_code": work_order.production_item,
            "quantity": work_order.qty,
            "work_order": work_order.name,
            "fg_item_qty": work_order.qty,
        })
    frappe.response['message'] = items
    return items

'''I am against this method: how will we know 
the specific work order to get the quantity from?'''

    
@frappe.whitelist()
def get_qty_from_first_work_order(work_order):
    """
    Fetch items and quantities from the first Work Order associated with the given Production Plan.
    Additionally, fetch items from the 'required_materials' child table where 'custom_item_group' is 'UPPER STOCK'.
    """
 
    work_order = frappe.get_doc("Work Order", work_order)
    required_materials = frappe.get_all(
        "Work Order Item",
        filters={
            "parent": work_order.name,
            "custom_item_group": "UPPER STOCK"
        },
        fields=["item_code"]
    )
    response = { "work_order_qty": work_order.qty, "upper_stock_items": [] }
    
    if required_materials:
        response["upper_stock_items"] = required_materials[0]["item_code"]
    frappe.response["message"] = response
    
    


